packages/dds/tree/src/feature-libraries/mapTreeCursor.ts

```diff
@@ -0,0 +1,280 @@
+/*!
+ * Copyright (c) Microsoft Corporation and contributors. All rights reserved.
+ * Licensed under the MIT License.
+ */
+
+import { assert } from "@fluidframework/common-utils";
+import {
+    ITreeCursorNew as ITreeCursor,
+    CursorLocationType,
+    mapCursorFieldNew as mapCursorField,
+} from "../forest";
+import {
+    FieldKey,
+    TreeType,
+    UpPath,
+    Value,
+    MapTree,
+    getMapTreeField,
+} from "../tree";
+import { fail } from "../util";
+
+/**
+ * @returns an ITreeCursor for a single MapTree.
+ */
+export function singleMapTreeCursor(root: MapTree): ITreeCursor {
+    return new MapCursor(root);
+}
+
+type SiblingsOrKey = readonly MapTree[] | readonly FieldKey[];
+
+/**
+ * An ITreeCursor implementation for MapTree.
+ *
+ * TODO:
+ * This is based off of TextCursor,
+ * and likely could be further optimized by taking a different approach using map iterators.
+ */
+class MapCursor implements ITreeCursor {
+    /**
+     * Indices traversed to visit this node: does not include current level (which is stored in `index`).
+     * Even indexes are of nodes and odd indexes are for fields.
+     */
+    private readonly indexStack: number[] = [];
+    /**
+     * Siblings into which indexStack indexes: does not include current level (which is stored in `siblings`).
+     * Even indexes are of nodes and odd indexes are for fields.
+     */
+    private readonly siblingStack: SiblingsOrKey[] = [];
+
+    private siblings: SiblingsOrKey;
+
+    /**
+     * Index into `siblings`.
+     */
+    private index: number;
+
+    /**
+     * Might start at special root where fields are detached sequences.
+     */
+    public constructor(root: MapTree) {
+        this.siblings = [root];
+        this.index = 0;
+    }
+
+    public getFieldKey(): FieldKey {
+        // assert(this.mode === CursorLocationType.Fields, "must be in fields mode");
```

Comment:
These are on a performance sensitive hot path and add overhead. Since we currently have no way to have asserts only in test or debug configurations, I just disabled them. Now that we have more perf tests, we might consider seeing if some of these are cheep enough to reenable.